File: Octave/cholesky.m
## Copyright (C) 2019 mherr
##
## This program is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see
## <https://www.gnu.org/licenses/>.

## -*- texinfo -*-
## @deftypefn {} {@var{retval} =} biseccion (@var{input1}, @var{input2})
##
## @seealso{}
## @end deftypefn

## Author: mherr <mherr@MSI>
## Created: 2019-08-11

## Metodo directo de factorizacion de Cholesky para solucion de sistemas de 
## ecuaciones.
## Entradas: Matriz de Coeficientes y matriz de términos independientes.
# Salidas: Solucion .
function [result] = cholesky(a, b)
    n = length(a);
    
    function [L] = get_L(a, n)
        L = zeros(n);
        for i = 1:n
            for j = 1:n
                if i == j
                    sum1 = 0;
                    for k = 1:j
                        sum1 += L(j, k)**2;
                    endfor
                    L(i, j) = sqrt(a(i, j) - sum1);
                else 
                    if i > j
                        sum2 = 0;
                        for k = 1:j
                            sum2 += L(i, k)*L(j, k);
                        endfor    
                        L(i, j) = (a(i, j) - sum2) / L(j, j);
                    end
                end
            endfor
        endfor    
    endfunction

    function [result] = sustitucion_hacia_atras(a, b, n)
        result = zeros(1, n);
        i = n;
        while i > 0
            resultTemp = b(i);
            j = n;
            while j > 0
                if i == j
                    result(j) = resultTemp/a(i,j);
                    break
                else
                    resultTemp -= a(i,j)*result(j);
                end
                j -= 1;
            endwhile    
            i -= 1;
        endwhile
    endfunction
    
    function [result] = sustitucion_hacia_adelante(a, b, n)
        result = zeros(1, n);
        for i = 1:n
            resultTemp = b(i);
            for j = 1:n
                if i == j
                    result(j) = resultTemp/a(i,j);
                    break
                else
                    resultTemp -= a(i,j)*result(j);
                end
            endfor
        endfor    
    endfunction
    L = get_L(a, n);
    y = sustitucion_hacia_adelante(L, b, n);
    result = sustitucion_hacia_atras(transpose(L), y, n);
endfunction    

File: Python/cholesky.py
from math import *
import numpy as np

def cholesky(a, b):
    n = len(a)
    L = get_L(a, b, n)
    Lt =[[L[j][i] for j in range(n)] for i in range(n)]
    y = sustitucion_hacia_adelante(L, b, n)
    return sustitucion_hacia_atras(Lt, y, n)

def get_L(a, b, n):
    L = [[0 for i in range(n)] for j in range(n)]

    for i in range(n):
        for j in range(n):
            if i == j:
                sum1 = 0
                for k in range(j):
                    sum1 += L[j][k]**2
                L[i][j] = sqrt(a[i][j] - sum1)
            elif i > j:
                sum2 = 0
                for k in range(j):
                    sum2 += L[i][k]*L[j][k]
                L[i][j] = (a[i][j] - sum2) / L[j][j]
    return L

def sustitucion_hacia_atras(a, b, n):
    result = [None]*n
    i = n - 1
    while i >= 0:
        resultTemp = b[i]
        j = n - 1
        while j >= 0:
            if i == j:
                result[j] = resultTemp/a[i][j]
                break
            else:
                resultTemp -= a[i][j]*result[j]
            j -= 1
        i -= 1
    return result


def sustitucion_hacia_adelante(a, b, n):
    result = [None]*n
    for i in range (n):
        resultTemp = b[i]
        for j in range(n):
            if i == j:
                result[j] = resultTemp/a[i][j]
                break
            else:
                resultTemp -= a[i][j]*result[j]
    return result


print(cholesky([[25, 15, -5, -10], [15, 10, 1, -7], [-5, 1, 21, 4], [-10, -7, 4, 18]], [-25, -19, -21, -5]))

